source/common/config/delta_subscription_state.cc

```diff
@@ -14,37 +14,80 @@ namespace Config {
       ttl_(
           [this](const auto& expired) {
             Protobuf::RepeatedPtrField<std::string> removed_resources;
             for (const auto& resource : expired) {
-              setResourceWaitingForServer(resource);
-              removed_resources.Add(std::string(resource));
+              if (auto maybe_resource = getRequestedResourceState(resource);
+                  maybe_resource.has_value()) {
+                maybe_resource->setAsWaitingForServer();
+                removed_resources.Add(std::string(resource));
+              } else if (auto erased_count = wildcard_resource_state_.erase(resource) +
+                                             ambiguous_resource_state_.erase(resource);
+                         erased_count > 0) {
+                removed_resources.Add(std::string(resource));
+              }
             }
 
             watch_map_.onConfigUpdate({}, removed_resources, "");
           },
           dispatcher, dispatcher.timeSource()),
-      type_url_(std::move(type_url)), wildcard_(wildcard), watch_map_(watch_map),
-      local_info_(local_info), dispatcher_(dispatcher) {}
+      type_url_(std::move(type_url)), watch_map_(watch_map), local_info_(local_info),
+      dispatcher_(dispatcher) {}
 
 void DeltaSubscriptionState::updateSubscriptionInterest(
     const absl::flat_hash_set<std::string>& cur_added,
     const absl::flat_hash_set<std::string>& cur_removed) {
   for (const auto& a : cur_added) {
-    setResourceWaitingForServer(a);
+    if (in_initial_legacy_wildcard_ && a != Wildcard) {
+      in_initial_legacy_wildcard_ = false;
+    }
+    // If the requested resource existed as a wildcard resource,
+    // transition it to requested. Otherwise mark it as a resource
+    // waiting for the server to receive the version.
+    if (auto it = wildcard_resource_state_.find(a); it != wildcard_resource_state_.end()) {
+      requested_resource_state_.insert_or_assign(a, ResourceState::withVersion(it->second));
+      wildcard_resource_state_.erase(it);
+    } else if (it = ambiguous_resource_state_.find(a); it != ambiguous_resource_state_.end()) {
+      requested_resource_state_.insert_or_assign(a, ResourceState::withVersion(it->second));
+      ambiguous_resource_state_.erase(it);
+    } else {
+      requested_resource_state_.insert_or_assign(a, ResourceState::waitingForServer());
+    }
+    ASSERT(requested_resource_state_.contains(a));
+    ASSERT(!wildcard_resource_state_.contains(a));
+    ASSERT(!ambiguous_resource_state_.contains(a));
     // If interest in a resource is removed-then-added (all before a discovery request
     // can be sent), we must treat it as a "new" addition: our user may have forgotten its
     // copy of the resource after instructing us to remove it, and need to be reminded of it.
     names_removed_.erase(a);
     names_added_.insert(a);
   }
   for (const auto& r : cur_removed) {
-    removeResourceState(r);
+    in_initial_legacy_wildcard_ = false;
```

Comment:
Hmm, What if we are in a legacy wildcard state and have none of the resources in cur_removed in `requested_resource_state_`?

Reply:
I'd say that legacy wildcard state lasts as long as we either do not update our subscription interest or our subscription interest is an empty set for both added and removed resources. I have made an exception in that we keep the legacy wildcard state when the subscription interest consist only of adding wildcard resource.

I'd risk saying that a scenario like that (being in legacy wildcard state and then removing some resource) is rather weird, but I'm not sure how to behave in this case. I suppose that in this case I shouldn't be adding anything to `names_removed_`, which could help in not sending any message that changes nothing.

Reply:
I changed it, so ineffectual unsubscriptions are basically no-ops, so it will preserve the legacy wildcard state.